sys/fmt/fmt.c

```diff
@@ -198,62 +198,65 @@ size_t fmt_s16_dec(char *out, int16_t val)
     return fmt_s32_dec(out, val);
 }
 
-size_t fmt_s16_dfp(char *out, int16_t val, unsigned fp_digits)
+size_t fmt_s16_dfp(char *out, int16_t val, int fp_digits)
 {
-    int16_t absolute, divider;
     size_t pos = 0;
-    size_t div_len, len;
-    unsigned e;
     char tmp[4];
 
-    if (fp_digits > 4) {
+    /* make sure inputs are valid */
+    if ((fp_digits > 6) || (fp_digits < -4) || (val == -32768)) {
         return 0;
     }
     if (fp_digits == 0) {
         return fmt_s16_dec(out, val);
     }
     if (val < 0) {
         if (out) {
-            out[pos++] = '-';
+            out[pos] = '-';
         }
+        ++pos;
         val *= -1;
     }
 
-    e = pwr(10, fp_digits);
-    absolute = (val / (int)e);
-    divider = val - (absolute * e);
-
-    pos += fmt_s16_dec(&out[pos], absolute);
-
-    if (!out) {
-        return pos + 1 + fp_digits;     /* abs len + decimal point + divider */
-    }
+    if (fp_digits < 0) {
+        fp_digits *= -1;
+        int e = (int)_tenmap[fp_digits];
```

Comment:
wait, actually this won't break: fp_digits is defined to be `<= 4`, hence the most we get for `e` is `10000`, which fits quite well into those 16-bit.